routing/payment_lifecycle.go

```diff
@@ -266,8 +266,8 @@ lifecycle:
 			return exitWithErr(err)
 		}
 
-		// NOTE: might cause an infinite loop, see notes in
-		// `requestRoute` for details.
+		// We may not be able to find a route. In that case, we
+		// continue the loop and move straight to the next iteration.
 		if rt == nil {
```

Comment:
Very good question. I think depends on how well the API design of `RequestRoute` really is. Ideally we should never return `nil, nil`, but cannot be sure here as I need to check more on the path finding code - logging an error now.

Reply:
This is what I mean:
```
 rlncli queryroutes  --dest 0215e87491363cc6dba86d29ee32fa29bcd2cd972ffce545583efd3cdfa58a2526  --amt 10000  | rlncli sendtoroute --payment_hash 7d4745d627e6bc5fecb388d83d09e6561f7bf6d1c7c9370edaf2cf013acb3ba7 --skip_temp_err  -
{
    "attempt_id":  "8001",
    "status":  "FAILED",
    "route":  {
        "total_time_lock":  568,
        "total_fees":  "1",
        "total_amt":  "10001",
        "hops":  [
            {
                "chan_id":  "211106232598528",
                "chan_capacity":  "3000000",
                "amt_to_forward":  "10000",
                "fee":  "1",
                "expiry":  488,
                "amt_to_forward_msat":  "10000000",
                "fee_msat":  "1010",
                "pub_key":  "0355402a07589a522375f83b6f6c42c4144d219ec3810e0a2aefb8ae99bdabd095",
                "tlv_payload":  true,
                "mpp_record":  null,
                "amp_record":  null,
                "custom_records":  {},
                "metadata":  "",
                "blinding_point":  "",
                "encrypted_data":  "",
                "total_amt_msat":  "0"
            },
            {
                "chan_id":  "407918813970433",
                "chan_capacity":  "250000",
                "amt_to_forward":  "10000",
                "fee":  "0",
                "expiry":  488,
                "amt_to_forward_msat":  "10000000",
                "fee_msat":  "0",
                "pub_key":  "0215e87491363cc6dba86d29ee32fa29bcd2cd972ffce545583efd3cdfa58a2526",
                "tlv_payload":  true,
                "mpp_record":  null,
                "amp_record":  null,
                "custom_records":  {},
                "metadata":  "",
                "blinding_point":  "",
                "encrypted_data":  "",
                "total_amt_msat":  "0"
            }
        ],
        "total_fees_msat":  "1010",
        "total_amt_msat":  "10001010"
    },
    "attempt_time_ns":  "1698065001317750000",
    "resolve_time_ns":  "1698065001638317000",
    "failure":  {
        "code":  "UNKNOWN_NEXT_PEER",
        "channel_update":  null,
        "htlc_msat":  "0",
        "onion_sha_256":  "",
        "cltv_expiry":  0,
        "flags":  0,
        "failure_source_index":  1,
        "height":  0
    },
    "preimage":  ""
}
```

Now trying to resolve the non-MPP payment will lead to:

```
rlncli queryroutes  --dest 0215e87491363cc6dba86d29ee32fa29bcd2cd972ffce545583efd3cdfa58a2526  --amt 10000  | rlncli sendtoroute --payment_hash 7d4745d627e6bc5fecb388d83d09e6561f7bf6d1c7c9370edaf2cf013acb3ba7 --skip_temp_err  -
[lncli] rpc error: code = AlreadyExists desc = payment is in transition
```

```
rlncli queryroutes  --dest 0215e87491363cc6dba86d29ee32fa29bcd2cd972ffce545583efd3cdfa58a2526  --amt 10000  | rlncli sendtoroute --payment_hash 7d4745d627e6bc5fecb388d83d09e6561f7bf6d1c7c9370edaf2cf013acb3ba7   -
[lncli] rpc error: code = AlreadyExists desc = payment is in transition
```

Deleteting this payment won't work because its inflight:

```
rlncli deletepayments --all
Removing failed payments, this might take a while...
{
	"ok": true
}
```
```
 rlncli listpayments --include_incomplete
  {
            "payment_hash":  "7d4745d627e6bc5fecb388d83d09e6561f7bf6d1c7c9370edaf2cf013acb3ba7",
            "value":  "10000",
            "creation_date":  "1698065001",
            "fee":  "0",
            "payment_preimage":  "0000000000000000000000000000000000000000000000000000000000000000",
            "value_sat":  "10000",
            "value_msat":  "10000000",
            "payment_request":  "",
            "status":  "IN_FLIGHT",
            "fee_sat":  "0",
            "fee_msat":  "0",
            "creation_time_ns":  "1698065001298205000",
            "htlcs":  [
                {
                    "attempt_id":  "8001",
                    "status":  "FAILED",
                    "route":  {
                        "total_time_lock":  568,
                        "total_fees":  "1",
                        "total_amt":  "10001",
                        "hops":  [
                            {
                                "chan_id":  "211106232598528",
                                "chan_capacity":  "3000000",
                                "amt_to_forward":  "10000",
                                "fee":  "1",
                                "expiry":  488,
                                "amt_to_forward_msat":  "10000000",
                                "fee_msat":  "1010",
                                "pub_key":  "0355402a07589a522375f83b6f6c42c4144d219ec3810e0a2aefb8ae99bdabd095",
                                "tlv_payload":  true,
                                "mpp_record":  null,
                                "amp_record":  null,
                                "custom_records":  {},
                                "metadata":  "",
                                "blinding_point":  "",
                                "encrypted_data":  "",
                                "total_amt_msat":  "0"
                            },
                            {
                                "chan_id":  "407918813970433",
                                "chan_capacity":  "250000",
                                "amt_to_forward":  "10000",
                                "fee":  "0",
                                "expiry":  488,
                                "amt_to_forward_msat":  "10000000",
                                "fee_msat":  "0",
                                "pub_key":  "0215e87491363cc6dba86d29ee32fa29bcd2cd972ffce545583efd3cdfa58a2526",
                                "tlv_payload":  true,
                                "mpp_record":  null,
                                "amp_record":  null,
                                "custom_records":  {},
                                "metadata":  "",
                                "blinding_point":  "",
                                "encrypted_data":  "",
                                "total_amt_msat":  "0"
                            }
                        ],
                        "total_fees_msat":  "1010",
                        "total_amt_msat":  "10001010"
                    },
                    "attempt_time_ns":  "1698065001317750000",
                    "resolve_time_ns":  "1698065001638317000",
                    "failure":  {
                        "code":  "UNKNOWN_NEXT_PEER",
                        "channel_update":  null,
                        "htlc_msat":  "0",
                        "onion_sha_256":  "",
                        "cltv_expiry":  0,
                        "flags":  0,
                        "failure_source_index":  1,
                        "height":  0
                    },
                    "preimage":  ""
                }
            ],
            "payment_index":  "9004",
            "failure_reason":  "FAILURE_REASON_NONE"
        }
    ],
```

Reply:
This can only be resolved by launching a payment with the MPP records set (hence MPP payment) which clears this payment from the db.

So I suggest, either forcing a timeout of the payment-lifecyle or forcing an MPP payment when the skip_temp_err is selected

Reply:
Thanks for the experiment! Yeah this totally makes sense, added a new commit to fix this case.